### PR TITLE
Object detection tf2 Yolov3 example

### PR DESCRIPTION
**Pull request by Le-Zheng**

### CODE REVIEW

**Comment by Le-Zheng** [comment]
jenkins passed: http://10.239.176.111:18888/job/ZOO-PR-Validation/5812/

File: pyzoo/zoo/examples/orca/learn/tf2/yolov3/yoloV3.py [hunk 1]
```diff
@@ -12,6 +13,26 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/tf2/yolov3/yoloV3.py:15]: Add this file is adapted from ....

File: pyzoo/zoo/examples/orca/learn/tf2/yolov3/yoloV3.py [hunk 2]
```diff
@@ -37,44 +58,47 @@
...
     anchor_idxs = tf.cast(anchor_idxs, tf.int32)
 
-    indexes = tf.TensorArray(tf.int32, 1, dynamic_size=True)
-    updates = tf.TensorArray(tf.float32, 1, dynamic_size=True)
-    idx = 0
-    for i in tf.range(N):
-        for j in tf.range(tf.shape(y_true)[1]):
-            if tf.equal(y_true[i][j][2], 0):
-                continue
+    def outer_comp(i):
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/tf2/yolov3/yoloV3.py:65]: Do we know why we need these changes? Is it due to `tf.distribute.Strategy`?

File: pyzoo/zoo/examples/orca/learn/tf2/yolov3/yoloV3.py [hunk 3]
```diff
@@ -0,0 +1,677 @@
...
+        return model
+
+    # prepare data
+    class_map = {name: idx for idx, name in enumerate(
+        open(options.names).read().splitlines())}
+    dataset_path = os.path.join(options.data_dir, "VOCdevkit")
+    voc_train_path = os.path.join(options.output_data, "train_dataset")
+    voc_val_path = os.path.join(options.output_data, "val_dataset")
+
+    write_voc(dataset_path, splits_names=[(options.data_year, options.split_name_train)],
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/tf2/yolov3/yoloV3.py:597]: change to `image.write_parquet(format=...)`

File: pyzoo/zoo/examples/orca/learn/tf2/yolov3/yoloV3.py [hunk 4]
```diff
@@ -0,0 +1,677 @@
...
+              classes=class_map)
+    write_voc(dataset_path, splits_names=[(options.data_year, options.split_name_test)],
+              output_path="file://" + voc_val_path,
+              classes=class_map)
+
+    output_types = {"image": tf.string, "label": tf.float32, "image_id": tf.string}
+    output_shapes = {"image": (), "label": (None, 5), "image_id": ()}
+
+    def train_data_creator(config, batch_size):
+        train_dataset = read_parquet("tf_dataset", input_path=voc_train_path,
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/tf2/yolov3/yoloV3.py:608]: change to `read_parquet(format="tf_dataset", path=...)`

File: pyzoo/zoo/examples/orca/learn/tf2/yolov3/yoloV3.py (hunk 2 above)
Comment (Le-Zheng) [reply pyzoo/zoo/examples/orca/learn/tf2/yolov3/yoloV3.py:65]: It seems TensorFlow autograph limitations on python code. I notice this issue later in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/autograph/g3doc/reference/limitations.md

File: pyzoo/zoo/examples/orca/learn/tf2/yolov3/yoloV3.py [hunk 5]
```diff
@@ -13,6 +13,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+#=============================================================================
+# This file is adapted from https://github.com/zzh8829/yolov3-tf2
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/tf2/yolov3/yoloV3.py:17]: which files? and then you need to add "Copyright (c) 2019 Zihao Zhang" below as in the original repo

**Comment by Le-Zheng** [comment]
jenkins passed: http://10.239.176.111:18888/job/ZOO-PR-Validation/5844/

**Comment by yangw1234** [comment]
I think maybe it is better to also post the training result in readme as well as some visualization of the predicted bounding boxes?

**Comment by Le-Zheng** [comment]
> I think maybe it is better to also post the training result in readme as well as some visualization of the predicted bounding boxes?

Ok. I will add that in another pull request.

**Pull request merged**